### PR TITLE
Push Tokens, Memory Leaks, ESLint

### PR DESCRIPTION
**Pull request by michaelkantor**
1. Adds APIs for registering and remove push tokens
2. ESLint was updated, lots of new code cleanup required.
3. Prevents destroyed queries from processing their async results
4. Memory Leak Fixes:
- Messages in a Query weren't destroyed when the Query was destroyed.
- Adds _tempConversationsHash, _tempMessagesHash and _tempId properties rather than leaving a floating pointer for 60 seconds that would show up in memory profiler.  Also allows for long term use of temporary IDs.
- _removeMessage replaced with destroy() for proper cleanup
- Conversation no longer destroy's lastMessage; destroying all Messages associated with a Conversation is already handled by Client
- part.on('parts:send') creates a persistent link between a MessagePart and its Message.  Replaced with once().
- MessageParts are no longer overridden by _populateFromServer due to not having an ID.  Now assigns an ID.

### CODE REVIEW

File: src/client-authenticator.js [hunk 1]
```diff
@@ -77,9 +77,11 @@ class Client extends Root {
     // We won't copy in userId; thats set from the identity-token... or from cache.
     // the userId argument is a way to identify if there has been a change of users.
     const requestedUserId = options.userId;
-    let cachedSessionData = '', cachedUserId = '';
+    let cachedSessionData = '',
+      cachedUserId = '';
```
Comment (nilgradisnik) [inline src/client-authenticator.js]: Let's avoid declaring variables using commas like that, especially with that indentation it just looks off.

Comment (michaelkantor) [inline src/client-authenticator.js]: I'm of the opinion that the ONLY time its OK to declare multiple variables on the same line is if your not assigning any values.  As soon as any assignments take place, single lines like that make it hard to find where things are declared.

File: src/client-authenticator.js [hunk 2]
```diff
@@ -538,6 +541,77 @@ class Client extends Root {
...
+   * For use with native code only (Cordova, React Native, Titanium, etc...)
+   *
+   * @method unregisterPushToken
+   * @param {string} deviceId - Your IOS device's device ID
+   * @param {Function} [callback=null] - Optional callback
+   * @param {layer.LayerError} callback.error - LayerError if there was an error; null if successful
+   */
+  unregisterPushToken(deviceId, callback) {
+    this.xhr({
+      url: 'push_tokens/' + deviceId,
```
Comment (nilgradisnik) [inline src/client-authenticator.js]: I'd encourage using es6 string templates.

File: src/client.js [hunk 3]
```diff
@@ -508,7 +516,12 @@ class Client extends ClientAuth {
    * @private
    */
   _triggerLogger(eventName, evt) {
-    if (['conversations:add', 'conversations:remove', 'conversations:change', 'messages:add', 'messages:remove', 'messages:change', 'challenge', 'ready'].indexOf(eventName) !== -1) {
+    const infoEvents = [
+      'conversations:add', 'conversations:remove',
+      'conversations:change', 'messages:add',
+      'messages:remove', 'messages:change',
+      'challenge', 'ready'];
```
Comment (nilgradisnik) [inline src/client.js]: Just a stylistic thing, but I'd encourage ending array `];` in a new line.

File: src/message.js [hunk 4]
```diff
@@ -706,10 +700,18 @@ class Message extends Syncable {
    * @param  {Object} m - Server description of the message
    */
   _populateFromServer(message) {
+    var id = this.id;
```
Comment (nilgradisnik) [inline src/message.js]: This should probably be `const`. And maybe rename this `id` to something more descriptive?

File: src/client-authenticator.js (hunk 1 above)
Comment (nilgradisnik) [inline src/client-authenticator.js]: Agreed. You are assigning an empty string value in this example.

Comment (michaelkantor) [inline src/client-authenticator.js]: This is a setting in eslint.  Changing this is part of a larger discussion... but eslint does not consider assignment to empty string to be different from Not an Assignment.

**Pull request merged**